core/server/proxy/src/main/java/alluxio/web/ProxyWebServer.java

```diff
@@ -72,13 +71,12 @@ public final class ProxyWebServer extends WebServer {
   private final RateLimiter mGlobalRateLimiter;
   private final FileSystem mFileSystem;
   private AsyncUserAccessAuditLogWriter mAsyncAuditLogWriter;
-  public static final String PROXY_S3_HANDLER_MAP = "Proxy S3 Handler Map";
-  public ConcurrentHashMap<Request, S3Handler> mS3HandlerMap = new ConcurrentHashMap<>();
+  public static final String S3_HANDLER_ATTRIBUTE = "Proxy S3 Handler Attribute";
 
   class ProxyListener implements HttpChannel.Listener {
     public void onComplete(Request request)
     {
-      S3Handler s3Hdlr = mS3HandlerMap.get(request);
+      S3Handler s3Hdlr = (S3Handler) request.getAttribute(S3_HANDLER_ATTRIBUTE);
```

Comment:
will it NPE when no attributes gets set?

Reply:
> will it NPE when no attributes gets set?

`S3Handler s3Hdlr = (S3Handler) null;` means s3Hdlr is null, like 'org.eclipse.jetty.server.Server#Server()'